File: WordPress/Classes/Models/Menu.m
#import "Menu.h"
#import "MenuItem.h"
#import "MenuLocation.h"
#import "Blog.h"

NSInteger const MenuDefaultID = -1;

@implementation Menu

@dynamic details;
@dynamic menuID;
@dynamic name;
@dynamic items;
@dynamic locations;
@dynamic blog;

+ (NSString *)entityName
{
    return NSStringFromClass([self class]);
}

+ (Menu *)newMenu:(NSManagedObjectContext *)managedObjectContext
{
    Menu *newMenu = [NSEntityDescription insertNewObjectForEntityForName:[Menu entityName] inManagedObjectContext:managedObjectContext];
    return newMenu;
}

+ (Menu *)defaultMenuForBlog:(Blog *)blog
{
    Menu *defaultMenu = nil;
    for (Menu *menu in blog.menus) {
        if (menu.menuID.integerValue == MenuDefaultID) {
            defaultMenu = menu;
            break;
        }
    }
    return defaultMenu;
}

+ (Menu *)newDefaultMenu:(NSManagedObjectContext *)managedObjectContext
{
    Menu *defaultMenu = [self newMenu:managedObjectContext];
    defaultMenu.menuID = @(MenuDefaultID);
    defaultMenu.name = [self defaultMenuName];
    return defaultMenu;
}

+ (NSString *)defaultMenuName
{
    return NSLocalizedString(@"Default Menu", @"Menu name for the defaut menu that is automatically generated.");
}

- (BOOL)isDefaultMenu
{
    return self.menuID.integerValue == MenuDefaultID;
}

@end

File: Modules/Sources/WordPressData/Swift/Menu.swift
import Foundation
import CoreData

@objc(Menu)
public class Menu: NSManagedObject {
    @objc public static let defaultID: Int = -1

    @objc(newMenu:)
    public static func newMenu(in managedObjectContext: NSManagedObjectContext) -> Menu {
        NSEntityDescription.insertNewObject(forEntityName: Menu.entityName(), into: managedObjectContext) as! Menu
    }

    @objc(defaultMenuForBlog:)
    public static func defaultMenu(for blog: Blog) -> Menu? {
        for menu in blog.menus ?? NSOrderedSet() {
            if let menu = menu as? Menu, menu.isDefaultMenu {
                return menu
            }
        }
        return nil
    }

    @objc(newDefaultMenu:)
    public static func newDefaultMenu(in managedObjectContext: NSManagedObjectContext) -> Menu {
        let defaultMenu = newMenu(in: managedObjectContext)
        defaultMenu.menuID = NSNumber(value: defaultID)
        defaultMenu.name = defaultMenuName()
        return defaultMenu
    }

    @objc(defaultMenuName)
    public static func defaultMenuName() -> String {
        return NSLocalizedString(
            "Default Menu",
            comment: "Menu name for the default menu that is automatically generated."
        )
    }

    @objc public var isDefaultMenu: Bool {
        return menuID?.intValue == Menu.defaultID
    }
}
